wardrobe/timestamp/src/lib.rs

```diff
@@ -0,0 +1,382 @@
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use core::marker::PhantomData;
+
+use parity_scale_codec::{Decode, Encode};
+use scale_info::TypeInfo;
+#[cfg(feature = "std")]
+use serde::{Deserialize, Serialize};
+use sp_core::H256;
+use sp_inherents::{CheckInherentsResult, InherentData};
+use sp_runtime::transaction_validity::TransactionPriority;
+use sp_std::{vec, vec::Vec};
+use sp_timestamp::InherentError::TooFarInFuture;
+use tuxedo_core::{
+    dynamic_typing::{DynamicallyTypedData, UtxoData},
+    ensure,
+    inherents::{TuxedoInherent, TuxedoInherentAdapter},
+    support_macros::{CloneNoBound, DebugNoBound, DefaultNoBound},
+    types::{Output, OutputRef, Transaction},
+    verifier::UpForGrabs,
+    ConstraintChecker, SimpleConstraintChecker, Verifier,
+};
+
+#[cfg(test)]
+mod cleanup_tests;
+#[cfg(test)]
+mod first_block_special_case_tests;
+#[cfg(test)]
+mod update_timestamp_tests;
+
+/// A piece-wide target for logging
+const LOG_TARGET: &str = "timestamp-piece";
+
+/// A timestamp, since the unix epocht, noted at some point in the history of the chain.
+/// It also records the block height in which it was included.
+#[derive(Debug, Encode, Decode, PartialEq, Eq, Clone, Copy, Default, PartialOrd, Ord)]
+pub struct Timestamp {
+    /// The time, in milliseconds, since the unix epoch.
+    pub time: u64,
+    /// The block number in which this timestamp was noted.
+    pub block: u32,
+}
+
+impl UtxoData for Timestamp {
+    const TYPE_ID: [u8; 4] = *b"time";
+}
+
+/// Options to configure the timestamp piece in your runtime.
+/// Currently we only need access to a block number.
+pub trait TimestampConfig {
+    /// A means of getting the current block height.
+    /// Probably this will be the Tuxedo Executive
+    fn block_height() -> u32;
+
+    /// The minimum amount of time by which the timestamp may be updated.
+    ///
+    /// The default is 2 seconds which should be slightly lower than most chains' block times.
+    const MINIMUM_TIME_INTERVAL: u64 = 2_000;
+
+    /// The maximum amount by which a valid block's timestamp may be ahead of an importing
+    /// node's current local time.
+    ///
+    /// Default is 1 minute.
+    const MAX_DRIFT: u64 = 60_000;
+
+    /// The minimum amount of time that must have passed before an old timestamp
+    /// may be cleaned up.
+    ///
+    /// Default is 1 day.
+    const MIN_TIME_BEFORE_CLEANUP: u64 = 1000 * 60 * 60 * 24;
+
+    /// The minimum number of blocks that must have passed before an old timestamp
+    /// may be cleaned up.
+    ///
+    /// Default is 15 thousand which is roughly equivalent to 1 day with 6 second
+    /// block times which is a common default in Substrate chains because of Polkadot.
+    const MIN_BLOCKS_BEFORE_CLEANUP: u32 = 15_000;
+}
+
+/// Reasons that setting or cleaning up the timestamp may go wrong.
+#[derive(Debug, Eq, PartialEq)]
+pub enum TimestampError {
+    /// UTXO data has an unexpected type
+    BadlyTyped,
+
```

Comment:
Extra space

Reply:
fixed in 6409334ff59118cda4dc786271e5e6ece396bd23